### ISSUE TITLE
Use of KeeWeb in enterprise with enforced settings

### ISSUE THREAD

**Issue by Nu11u5** [keeweb/keeweb#1609]
We are considering using KeeWeb in an enterprise environment but would need to be able to enforce certain settings such as locking the password database.  I know that when self-hosting, a config.js file can be specified, however as I understand this only changes the default settings and does no enforcement.

- Is there a way to enforce settings so they cannot be changed?
- Is there a way to use an alternative default configuration on the desktop application, similar to config.js as with self-hosted?

If this is not an option how viable would it be for a feature request?

**Comment by antelle** [comment]
Hi!

You're right, now `config.js` and `kw-config` meta-tag provide default settings, they can't be enforced this way. So, the answer to both questions is "No".

There's a possibility to provide settings via [`kw-config`](https://github.com/keeweb/keeweb/wiki/Configuration#json-app-config) meta tag, settings from there will be loaded always without any url parameters, but users can still modify it.

It's possible to implement settings enforcement, however, there's an important catch: KeeWeb is just another client for reading KDBX files. What users can do, is run KeePass, KeePassX, or even another instance of KeeWeb (save the HTML and open it in browser) without the config, load the KDBX file, and do whatever.

With this understanding, do you still see settings enforcement as something helpful?

**Comment by Nu11u5** [comment]
Yes, it's understood that another unmanaged app would not have the same restriction.  We just need to make sure that password management, as-offered, meets some basic requirements.  Currently our enterprise uses the mainline KeePass app on Windows, which has a "KeePass.config.enforced.xml" file.  However, to support other platforms we would need to use other KeePass apps - supporting and maintaining those would be a increasingly burdensome.  Also, KeePass needs several plugins (some which are not maintained as well) to match the features of KeeWeb.  So, KeeWeb is looking pretty good for unifying our platforms.

Would this feature also work on the desktop app?

**Comment by antelle** [comment]
It can be supported on the desktop app as well.

The way I see settings enforcement, there are two parts of it:

1. Settings application. The config should contain a new property that will indicate which settings can't be modified. If a setting is there, corresponding UI controls should be disabled, as well as the JS settings modification API should return an error.

2. Config delivery. Ideally it should be done in several ways:
    - our current `kw-config` meta-tag will just work, however it's a bit tricky to modify it;
    - there should be a possibility to make `config.js` loading mandatory, either with an HTTP header, or with another meta tag;
    - the desktop version can support a config file with enforced rules stored in a location not writable by the current user, something like next to the application binary or so.
